### ISSUE TITLE
Training steps vs training batches

### ISSUE THREAD

**Issue by del2z** [PyTorchLightning/pytorch-lightning#6984]
I have some doubts about training steps and training batches.
When training a model using max_steps, the number of training batches showed in progressing bar using 1 GPU is the same as using 8 GPUs. If using X GPUs, is the actual batch size during one optimization step X-times larger, which results in a smaller number of training batches?

The batch size should be expanded in multi-GPU setting according to DDP of pytorch (PyTorch Distributed: Experiences on Accelerating Data Parallel Training). From my observation in 1, the number of training batches equals to total number of steps (I don't think they should be the same). I wonder the DDP training mechanisms between pytorch and pytorch-lightning are still the same.

In pytorch, each device gets a copy of the model and other states exactly like the original one. After backwards, gradients from all devices are accumulated together and broadcasted to every device. Then models are updated using same gradients. Such process acts like splitting a large batch, execute forward and backward pass on a smaller, splitted part in each GPU and then accumulate all gradients and update model parameters.
It seems like pytorch-lightning treats a single optimization step in X GPUs as X steps. If so, the number of training batches becomes the number of steps. Only difference is that model parameters are updated once, not X times as expected. Another coincidence is that learning rate scheduler setting interval='step' updates learning rate in every batch.

I hope anyone could explain this.

**Comment by del2z** [comment]
python version: 3.8.5
pytorch version: 1.8.1
pytorch-lightning version: 1.2.5
OS version: CentOS 8

**Comment by awaelchli** [comment]
> When training a model using max_steps, the number of training batches showed in progressing bar using 1 GPU is the same as using 8 GPUs.

Because we show the progress bar on rank 0 only

>  I wonder the DDP training mechanisms between pytorch and pytorch-lightning are still the same.

yes they are! (they should be, or it's a bug)

> Such process acts like splitting a large batch, execute forward and backward pass on a smaller, splitted part in each GPU and then accumulate all gradients and update model parameters.

no, you are describing DP, not DDP. In DDP we don't split the batches by number of GPUs. 

> The batch size should be expanded in multi-GPU setting according to DDP of pytorch (PyTorch Distributed: Experiences on Accelerating Data Parallel Training)

Could you link to the article, please? I'd like to take a look.


Could you take our example and run it. 

`python pl_examples/basic_examples/simple_image_clasifier.py --accelerator ddp --gpus 2`

and let me know which observations / parameter settings you are not agreeing with?

**Comment by del2z** [comment]
Thanks for quick reply.
I understand the difference between DP and DDP. What I really mean is that, each GPU processes a batch from a distributed sampler, after all GPUs finish a batch, the gradients are then accumulated. For example, a dataloader with distributed sampler generates a batch of size 16, 8 GPUs can process 8 * 16 instances in one step, and then the gradients of these 128 instances are averaged to update model parameters. It's just like virtually splitting a batch of size 128 into 8 parts, processing all parts using 8 GPUs and merging their results. I suppose the resulting model trained by 8 GPUs with a batch size 16 should be same as the one trained by 1 GPU with batch size 128, regardless of randomness in sampling and ordering of instances.
> > Such process acts like splitting a large batch, execute forward and backward pass on a smaller, splitted part in each GPU and then accumulate all gradients and update model parameters.
> 
> no, you are describing DP, not DDP. In DDP we don't split the batches by number of GPUs.
> 
> > The batch size should be expanded in multi-GPU setting according to DDP of pytorch (PyTorch Distributed: Experiences on Accelerating Data Parallel Training)
The working mechanism of DDP can be found in this paper [1](https://arxiv.org/abs/2006.15704).

 I use following command to test the output of pytorch-lightning.
```sh
python ./simple_image_classifier.py \
    --num_nodes 1 \
    --accelerator 'ddp' \
    --gpus 2 \
    --batch_size 16 \
    --num_workers 1 \
    --max_epochs 2 \
    --accumulate_grad_batches 1 \
    --check_val_every_n_epoch 1 \
    --log_every_n_steps 1 \
    --flush_logs_every_n_steps 1 \
    --progress_bar_refresh_rate 1 \
    --auto_lr_find false \
    --replace_sampler_ddp false
```
Only 1 GPU:
![Screen Shot 2021-04-14 at 09 47 55](https://user-images.githubusercontent.com/14818752/114642439-80ba7500-9d06-11eb-96c4-54bbf2230e3a.png)
2 GPUs: 
![Screen Shot 2021-04-14 at 09 50 18](https://user-images.githubusercontent.com/14818752/114642640-d727b380-9d06-11eb-873b-a7db9f6e572e.png)
As you were saying, the progress bar only shows for rank 0. MNIST has 55000 images in the training set, all `3438 = ceil(55000 / 16)` batches are consumed by rank 0 GPU for only one GPU setting. However, in the later setting using 2 GPUs, rank 0 GPU only processes 1719 batches, which is not consistent with the output of progress bar. I got similar outputs when testing with even more GPUs. Why is that?

Further, I use `max_steps=900` instead of `max_epochs` and 8 GPUs. I have such a result.
![Screen Shot 2021-04-14 at 10 02 56](https://user-images.githubusercontent.com/14818752/114643617-9a5cbc00-9d08-11eb-9b97-16c7693c1c16.png)
The training stops at 900 batches, not 900 steps. Indeed, the training number is expected to be `900 * 16 * 8 = 115200` images, approximately 2.09 epochs.

It seems like pytorch-lightning regards a step as a batch, which I think is a misunderstanding. A step should be one duration of a whole update of model parameters accomplished in `optimizer.step()` call. In such sense, a steps may contain multiple batches.

**Comment by del2z** [comment]
After carefully inspect the source code, I found the mistake made in the launch command. `--replace_sampler_ddp false` should not be set in this case. Enabling `--replace_sampler_ddp` gives me the expected results.
Thanks again.
